Aarondonahue/io packed arrays

AaronDonahue opened the pull request:
Introduces infrastructure to support packed arrays for reading and writing from the field manager.

This commit expands the io_test to include a packed array and makes sure that output written and input read doesn't accidentally capture the padded values in the array.

This commit also cleans up a bunch of old unused I/O routines.

Addresses Issue #787 , which is also listed in #880 

[BFB]

E3SM-Autotester commented:
Status Flag 'Pre-Test Inspection' - Auto Inspected - Inspection Is Not Necessary for this Pull Request.

E3SM-Autotester commented:
Status Flag 'Pull Request AutoTester' - Testing Jenkins Projects:

<details>
<summary>
  <b>Pull Request Auto Testing STARTING</b> (click to expand)
</summary>

<h2>Build Information</h2>
<h3>Test Name: SCREAM_PullRequest_Autotester_Blake</h3>
<ul>
  <li>Build Num: 1022</li>
  <li>Status: STARTED</li>
</ul>

<h4>Jenkins Parameters</h4>
<table>
  <tr>
    <th>Parameter Name</th>
    <th>Value</th>
  </tr>

  <tr>
    <td>PR_LABELS</td>
    <td>AT: AUTOMERGE;B4B;Development Task;I/O</td>
  </tr>

  <tr>
    <td>PULLREQUESTNUM</td>
    <td>885</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_BRANCH</td>
    <td>aarondonahue/io_packed_arrays</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_SHA</td>
    <td>e1eeecf498bb88167b8d5781abdfb74cec3650f1</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_BRANCH</td>
    <td>master</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_SHA</td>
    <td>f4a43e7699293f01f0421850e6a1da4aca71c405</td>
  </tr>

  <tr>
    <td>TEST_REPO_ALIAS</td>
    <td>SCREAM</td>
  </tr>
</table>

<h2>Build Information</h2>
<h3>Test Name: SCREAM_PullRequest_Autotester_Weaver</h3>
<ul>
  <li>Build Num: 1009</li>
  <li>Status: STARTED</li>
</ul>

<h4>Jenkins Parameters</h4>
<table>
  <tr>
    <th>Parameter Name</th>
    <th>Value</th>
  </tr>

  <tr>
    <td>PR_LABELS</td>
    <td>AT: AUTOMERGE;B4B;Development Task;I/O</td>
  </tr>

  <tr>
    <td>PULLREQUESTNUM</td>
    <td>885</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_BRANCH</td>
    <td>aarondonahue/io_packed_arrays</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_SHA</td>
    <td>e1eeecf498bb88167b8d5781abdfb74cec3650f1</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_BRANCH</td>
    <td>master</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_SHA</td>
    <td>f4a43e7699293f01f0421850e6a1da4aca71c405</td>
  </tr>

  <tr>
    <td>TEST_REPO_ALIAS</td>
    <td>SCREAM</td>
  </tr>
</table>

<h2>Build Information</h2>
<h3>Test Name: SCREAM_PullRequest_Autotester_Mappy</h3>
<ul>
  <li>Build Num: 555</li>
  <li>Status: STARTED</li>
</ul>

<h4>Jenkins Parameters</h4>
<table>
  <tr>
    <th>Parameter Name</th>
    <th>Value</th>
  </tr>

  <tr>
    <td>PR_LABELS</td>
    <td>AT: AUTOMERGE;B4B;Development Task;I/O</td>
  </tr>

  <tr>
    <td>PULLREQUESTNUM</td>
    <td>885</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_BRANCH</td>
    <td>aarondonahue/io_packed_arrays</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_SHA</td>
    <td>e1eeecf498bb88167b8d5781abdfb74cec3650f1</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_BRANCH</td>
    <td>master</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_SHA</td>
    <td>f4a43e7699293f01f0421850e6a1da4aca71c405</td>
  </tr>

  <tr>
    <td>TEST_REPO_ALIAS</td>
    <td>SCREAM</td>
  </tr>
</table>

<h2>Using Repos:</h2>

<dl>
  <dt>Repo: SCREAM (E3SM-Project/scream)</dt>
  <ul>
    <li>Branch: aarondonahue/io_packed_arrays</li>
    <li>SHA: e1eeecf498bb88167b8d5781abdfb74cec3650f1</li>
    <li>Mode: TEST_REPO</li>
  </ul>
</dl>

Pull Request Author: AaronDonahue
</details>

E3SM-Autotester commented:
Status Flag 'Pull Request AutoTester' - Jenkins Testing: all Jobs PASSED

<details>
<summary>
  <b>Pull Request Auto Testing has PASSED</b> (click to expand)
</summary>

<h2>Build Information</h2>
<h3>Test Name: SCREAM_PullRequest_Autotester_Blake</h3>
<ul>
  <li>Build Num: 1022</li>
  <li>Status: PASSED</li>
</ul>

<h4>Jenkins Parameters</h4>
<table>
  <tr>
    <th>Parameter Name</th>
    <th>Value</th>
  </tr>

  <tr>
    <td>PR_LABELS</td>
    <td>AT: AUTOMERGE;B4B;Development Task;I/O</td>
  </tr>

  <tr>
    <td>PULLREQUESTNUM</td>
    <td>885</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_BRANCH</td>
    <td>aarondonahue/io_packed_arrays</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_SHA</td>
    <td>e1eeecf498bb88167b8d5781abdfb74cec3650f1</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_BRANCH</td>
    <td>master</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_SHA</td>
    <td>f4a43e7699293f01f0421850e6a1da4aca71c405</td>
  </tr>

  <tr>
    <td>TEST_REPO_ALIAS</td>
    <td>SCREAM</td>
  </tr>
</table>

<h2>Build Information</h2>
<h3>Test Name: SCREAM_PullRequest_Autotester_Weaver</h3>
<ul>
  <li>Build Num: 1009</li>
  <li>Status: PASSED</li>
</ul>

<h4>Jenkins Parameters</h4>
<table>
  <tr>
    <th>Parameter Name</th>
    <th>Value</th>
  </tr>

  <tr>
    <td>PR_LABELS</td>
    <td>AT: AUTOMERGE;B4B;Development Task;I/O</td>
  </tr>

  <tr>
    <td>PULLREQUESTNUM</td>
    <td>885</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_BRANCH</td>
    <td>aarondonahue/io_packed_arrays</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_SHA</td>
    <td>e1eeecf498bb88167b8d5781abdfb74cec3650f1</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_BRANCH</td>
    <td>master</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_SHA</td>
    <td>f4a43e7699293f01f0421850e6a1da4aca71c405</td>
  </tr>

  <tr>
    <td>TEST_REPO_ALIAS</td>
    <td>SCREAM</td>
  </tr>
</table>

<h2>Build Information</h2>
<h3>Test Name: SCREAM_PullRequest_Autotester_Mappy</h3>
<ul>
  <li>Build Num: 555</li>
  <li>Status: PASSED</li>
</ul>

<h4>Jenkins Parameters</h4>
<table>
  <tr>
    <th>Parameter Name</th>
    <th>Value</th>
  </tr>

  <tr>
    <td>PR_LABELS</td>
    <td>AT: AUTOMERGE;B4B;Development Task;I/O</td>
  </tr>

  <tr>
    <td>PULLREQUESTNUM</td>
    <td>885</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_BRANCH</td>
    <td>aarondonahue/io_packed_arrays</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_SHA</td>
    <td>e1eeecf498bb88167b8d5781abdfb74cec3650f1</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_BRANCH</td>
    <td>master</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_SHA</td>
    <td>f4a43e7699293f01f0421850e6a1da4aca71c405</td>
  </tr>

  <tr>
    <td>TEST_REPO_ALIAS</td>
    <td>SCREAM</td>
  </tr>
</table>
</details>

E3SM-Autotester commented:
Status Flag 'Pre-Merge Inspection' - - This Pull Request Requires Inspection... The code must be inspected by a member of the Team before Testing/Merging
           NO REVIEWS HAVE BEEN PERFORMED ON THIS PULL REQUEST!

E3SM-Autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

bartgol commented on components/scream/src/share/io/scream_scorpio_interface.cpp:
```diff
@@ -150,58 +176,39 @@ void grid_read_data_array(const std::string &filename, const std::string &varnam
+  {
+    grid_write_data_array_c2f_real_1d(filename.c_str(),varname.c_str(),dim_length,hbuf);
+  } else {
+    std::vector<Real> hbuf_new;
```
Since this chunk of code is the same as in the `read` routine, I think it's best to factor it out into an independent function. Something like
```
void remove_padding (const Real* src, Real* dst, const std::vector<int>& phys_dims, const int padding);
```
and then call
```
int dim_length = std::accumulate(dims.begin(),dims.end(),1,std::multiplies<int>());
std::vector<Real> hbuf_new(dim_length);
remove_padding(hbuf, hbuf_new.data(),dims,pack_extra);
```

bartgol commented on components/scream/src/share/io/scream_scorpio_interface.cpp:
```diff
@@ -138,9 +138,35 @@ void eam_pio_enddef(const std::string &filename) {
+      }
+      loc += pack_extra;
+    }
+    EKAT_REQUIRE_MSG((dim_length == dloc),"Error read input for field " + varname + ", packed and unpacked data variables don't match in size.");
```
Btw, since `dim_length=dims[0]*...*dims[n]`, you could do
```
auto slow_dim_len = dim_length / dims.back();
```

bartgol commented on components/scream/src/share/io/tests/io.cpp:
```diff
@@ -76,6 +76,20 @@ TEST_CASE("input_output_basic","io")
+  {
+    auto f_dev  = field_repo->get_field(it,"Physics").get_view();
+    auto f_host = Kokkos::create_mirror_view( f_dev );
+    Kokkos::deep_copy(f_host,f_dev);
```
You could simply do
```
Kokkos::deep_copy(f_dev,std::nan(""));
```
The `deep_copy` function is overloaded, and can either copy a src view into a dst view or a single value into a dst view.

bartgol commented on components/scream/src/share/io/tests/io.cpp:
```diff
@@ -195,36 +213,56 @@ std::shared_ptr<FieldRepository<Real>> get_test_repo(const Int num_lcols, const
+  auto field4 = repo->get_field(fid4);
+  auto fid4_extent = field4.get_header().get_alloc_properties().get_last_extent();
+  auto fid4_final_dim = field4.get_header().get_identifier().get_layout().dims().back();
+  REQUIRE(fid4_extent-fid4_final_dim > 0);
```
I think you need `>=0`. If pack sizes divides the field last dim, then `fid4_extent` and `fid4_final_dim` are the same.  In your case, you have nlevs=3 and a pack size 2, so it's fine. But if someone edits the test, it might be hard to catch what's going on.

You could make the test robust to edits, like this:
```
// Make sure that if ppl alter the test inputs, they see an error
EKAT_REQUIRE_MSG((last_dim % packsize)==0,
    "Error! Please, don't alter "nlevs" or "packsize" for this test.\n);
REQUIRE (extent-last_dim>0);
```

AaronDonahue replied:
you bring up a good point.  What happens when packsize = 1?  Shouldn't extent = last_dim?

AaronDonahue replied:
if so, then I think the REQUIRE shouldn't be there.  I'm essentially trying to avoid someone nullifying the packed test by accidentally making everything fit perfectly with nlev and packsize.

bartgol commented on components/scream/src/share/io/tests/io.cpp:
```diff
@@ -195,36 +213,56 @@ std::shared_ptr<FieldRepository<Real>> get_test_repo(const Int num_lcols, const
     {
       f2_hst(jj) = (jj+1)/10.0;
       f3_hst(ii,jj) = (ii) + (jj+1)/10.0;
+//      f4_hst(ii,jj) = (ii) + (jj+1)/10.0;
```
Remove commented line.

E3SM-Autotester commented:
Status Flag 'Pre-Merge Inspection' - - This Pull Request Requires Inspection... The code must be inspected by a member of the Team before Testing/Merging
           THE LAST COMMIT TO THIS PULL REQUEST HAS BEEN REVIEWED, BUT NOT ACCEPTED OR REQUIRES CHANGES

E3SM-Autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

AaronDonahue commented:
@bartgol , I think I addressed all of your comments in this latest commit.  Except one about checking that the packed field in io.cpp is actually a padded field.  I responded to your comment with a question, hopefully we can figure out the right away to put in that check.

I also took a stab at adding the `get_padding` function to field allocation properties.  Can you take a look at that?

E3SM-Autotester commented:
Status Flag 'Pre-Test Inspection' - Auto Inspected - Inspection Is Not Necessary for this Pull Request.

E3SM-Autotester commented:
Status Flag 'Pull Request AutoTester' - Testing Jenkins Projects:

<details>
<summary>
  <b>Pull Request Auto Testing STARTING</b> (click to expand)
</summary>

<h2>Build Information</h2>
<h3>Test Name: SCREAM_PullRequest_Autotester_Blake</h3>
<ul>
  <li>Build Num: 1032</li>
  <li>Status: STARTED</li>
</ul>

<h4>Jenkins Parameters</h4>
<table>
  <tr>
    <th>Parameter Name</th>
    <th>Value</th>
  </tr>

  <tr>
    <td>PR_LABELS</td>
    <td>AT: AUTOMERGE;B4B;Development Task;I/O</td>
  </tr>

  <tr>
    <td>PULLREQUESTNUM</td>
    <td>885</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_BRANCH</td>
    <td>aarondonahue/io_packed_arrays</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_SHA</td>
    <td>ada8570a0a0351df5232ce8942aec23c62581859</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_BRANCH</td>
    <td>master</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_SHA</td>
    <td>f4a43e7699293f01f0421850e6a1da4aca71c405</td>
  </tr>

  <tr>
    <td>TEST_REPO_ALIAS</td>
    <td>SCREAM</td>
  </tr>
</table>

<h2>Build Information</h2>
<h3>Test Name: SCREAM_PullRequest_Autotester_Weaver</h3>
<ul>
  <li>Build Num: 1019</li>
  <li>Status: STARTED</li>
</ul>

<h4>Jenkins Parameters</h4>
<table>
  <tr>
    <th>Parameter Name</th>
    <th>Value</th>
  </tr>

  <tr>
    <td>PR_LABELS</td>
    <td>AT: AUTOMERGE;B4B;Development Task;I/O</td>
  </tr>

  <tr>
    <td>PULLREQUESTNUM</td>
    <td>885</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_BRANCH</td>
    <td>aarondonahue/io_packed_arrays</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_SHA</td>
    <td>ada8570a0a0351df5232ce8942aec23c62581859</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_BRANCH</td>
    <td>master</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_SHA</td>
    <td>f4a43e7699293f01f0421850e6a1da4aca71c405</td>
  </tr>

  <tr>
    <td>TEST_REPO_ALIAS</td>
    <td>SCREAM</td>
  </tr>
</table>

<h2>Build Information</h2>
<h3>Test Name: SCREAM_PullRequest_Autotester_Mappy</h3>
<ul>
  <li>Build Num: 565</li>
  <li>Status: STARTED</li>
</ul>

<h4>Jenkins Parameters</h4>
<table>
  <tr>
    <th>Parameter Name</th>
    <th>Value</th>
  </tr>

  <tr>
    <td>PR_LABELS</td>
    <td>AT: AUTOMERGE;B4B;Development Task;I/O</td>
  </tr>

  <tr>
    <td>PULLREQUESTNUM</td>
    <td>885</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_BRANCH</td>
    <td>aarondonahue/io_packed_arrays</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_SHA</td>
    <td>ada8570a0a0351df5232ce8942aec23c62581859</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_BRANCH</td>
    <td>master</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_SHA</td>
    <td>f4a43e7699293f01f0421850e6a1da4aca71c405</td>
  </tr>

  <tr>
    <td>TEST_REPO_ALIAS</td>
    <td>SCREAM</td>
  </tr>
</table>

<h2>Using Repos:</h2>

<dl>
  <dt>Repo: SCREAM (E3SM-Project/scream)</dt>
  <ul>
    <li>Branch: aarondonahue/io_packed_arrays</li>
    <li>SHA: ada8570a0a0351df5232ce8942aec23c62581859</li>
    <li>Mode: TEST_REPO</li>
  </ul>
</dl>

Pull Request Author: AaronDonahue
</details>

E3SM-Autotester commented:
Status Flag 'Pull Request AutoTester' - Jenkins Testing: all Jobs PASSED

<details>
<summary>
  <b>Pull Request Auto Testing has PASSED</b> (click to expand)
</summary>

<h2>Build Information</h2>
<h3>Test Name: SCREAM_PullRequest_Autotester_Blake</h3>
<ul>
  <li>Build Num: 1032</li>
  <li>Status: PASSED</li>
</ul>

<h4>Jenkins Parameters</h4>
<table>
  <tr>
    <th>Parameter Name</th>
    <th>Value</th>
  </tr>

  <tr>
    <td>PR_LABELS</td>
    <td>AT: AUTOMERGE;B4B;Development Task;I/O</td>
  </tr>

  <tr>
    <td>PULLREQUESTNUM</td>
    <td>885</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_BRANCH</td>
    <td>aarondonahue/io_packed_arrays</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_SHA</td>
    <td>ada8570a0a0351df5232ce8942aec23c62581859</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_BRANCH</td>
    <td>master</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_SHA</td>
    <td>f4a43e7699293f01f0421850e6a1da4aca71c405</td>
  </tr>

  <tr>
    <td>TEST_REPO_ALIAS</td>
    <td>SCREAM</td>
  </tr>
</table>

<h2>Build Information</h2>
<h3>Test Name: SCREAM_PullRequest_Autotester_Weaver</h3>
<ul>
  <li>Build Num: 1019</li>
  <li>Status: PASSED</li>
</ul>

<h4>Jenkins Parameters</h4>
<table>
  <tr>
    <th>Parameter Name</th>
    <th>Value</th>
  </tr>

  <tr>
    <td>PR_LABELS</td>
    <td>AT: AUTOMERGE;B4B;Development Task;I/O</td>
  </tr>

  <tr>
    <td>PULLREQUESTNUM</td>
    <td>885</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_BRANCH</td>
    <td>aarondonahue/io_packed_arrays</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_SHA</td>
    <td>ada8570a0a0351df5232ce8942aec23c62581859</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_BRANCH</td>
    <td>master</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_SHA</td>
    <td>f4a43e7699293f01f0421850e6a1da4aca71c405</td>
  </tr>

  <tr>
    <td>TEST_REPO_ALIAS</td>
    <td>SCREAM</td>
  </tr>
</table>

<h2>Build Information</h2>
<h3>Test Name: SCREAM_PullRequest_Autotester_Mappy</h3>
<ul>
  <li>Build Num: 565</li>
  <li>Status: PASSED</li>
</ul>

<h4>Jenkins Parameters</h4>
<table>
  <tr>
    <th>Parameter Name</th>
    <th>Value</th>
  </tr>

  <tr>
    <td>PR_LABELS</td>
    <td>AT: AUTOMERGE;B4B;Development Task;I/O</td>
  </tr>

  <tr>
    <td>PULLREQUESTNUM</td>
    <td>885</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_BRANCH</td>
    <td>aarondonahue/io_packed_arrays</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_SOURCE_SHA</td>
    <td>ada8570a0a0351df5232ce8942aec23c62581859</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_BRANCH</td>
    <td>master</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_REPO</td>
    <td>https://github.com/E3SM-Project/scream</td>
  </tr>

  <tr>
    <td>SCREAM_TARGET_SHA</td>
    <td>f4a43e7699293f01f0421850e6a1da4aca71c405</td>
  </tr>

  <tr>
    <td>TEST_REPO_ALIAS</td>
    <td>SCREAM</td>
  </tr>
</table>
</details>

E3SM-Autotester commented:
Status Flag 'Pre-Merge Inspection' - - This Pull Request Requires Inspection... The code must be inspected by a member of the Team before Testing/Merging
           THE LAST COMMIT TO THIS PULL REQUEST HAS NOT BEEN REVIEWED YET!

E3SM-Autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

E3SM-Autotester commented:
Status Flag 'Pre-Merge Inspection' - SUCCESS: The last commit to this Pull Request has been INSPECTED AND APPROVED by [ bartgol ]!

E3SM-Autotester commented:
Status Flag 'Pull Request AutoTester' - Pull Request will be Automerged

E3SM-Autotester commented:
Merge on Pull Request# 885: IS A SUCCESS - Pull Request successfully merged
